### ISSUE TITLE
Long runtime for the quick algorithm on some large data structures

### ISSUE THREAD

**Issue by sigvesn** [juji-io/editscript#29]
At microdata.no we use editscript with the `:quick` diff algorithm to generate diff's between potentially large edn structures.

This usually works very well, but i have encountered a special case where the `editscript/diff` with the `:quick` algorithm is very slow.

It is hard to pinpoint exactly what it is in the data that causes this, as the compared files are about 2.4MB each, but here is a repo with two example files (which also contains code to run the diff):

https://github.com/sigvesn/editscript-quickdiff-example

- each is a vector of about 200 maps
- the map entries potentially contains a lot of data
-  almost each map in the vectors are equal, except for one boolean field `:active` with is present for all maps in `old` but only one map in `new`  

`diff` using either the "non-quick" algorithm (or removing the `:active` field from the maps in `old`) runs in ~0.2-1 second

Using `:quick true` results in a runtime of ~7.5 minutes

**Comment by huahaiy** [comment]
Interesting, I will take a look.

**Comment by sigvesn** [comment]
Can the problem be connected to the fact that editscript seems to not be able to generate a diff at all, but instead returns the whole file as a replace action?

This is seen in the example output file from the diff generated by https://github.com/sigvesn/editscript-quickdiff-example/blob/main/src/main.clj#L19 at https://github.com/sigvesn/editscript-quickdiff-example/blob/main/out/diff.json 

I compared this against a diff created by a different tool: https://www.npmjs.com/package/small-json-diff created for the same files: https://github.com/sigvesn/editscript-quickdiff-example/blob/main/out/diff_small_json.json . This diff shows the expected changes as i described above

**Comment by huahaiy** [comment]
I can tell you what I know so far:

* The quick algorithm is extremely slow diffing two huge strings, because you enabled `:str-diff?`.  I am debugging this.
* The A* star algorithm chose to replace the whole thing, because the diff size is smaller that way. Our definition of diff size is strictly about the object count, replacing the whole thing is often smaller than having a diff data structure.

**Comment by huahaiy** [comment]
There's not much I can do at this point. The quick algorithm compare things blindly so it attempted to compare two huge strings, which is going to cost a lot of time. The A* algorithm did the right thing, but it's not the thing you would like. You will have to use something else. Closing.

**Comment by huahaiy** [comment]
On a second thought, I can introduce a time-out for string diff, so we can skip the computation if time is out.

**Comment by sigvesn** [comment]
Not using string diffing seems like a good solution for us. A time-out would also be welcome. 

Thank you very much for the help and advice!

**Comment by huahaiy** [comment]
0.6.0 now allow control string diffing, and added a timeout option.

**Comment by sigvesn** [comment]
The new :str-diff options and timeout seems to be working great. Some small notices:

- Using an `opts` map with `diff` no longer works in clojure 1.10 and will throw an `java.lang.IllegalArgumentException: No value supplied for key: {}` error. Might be worth to note that 1.11 is required.

- Small error in the [changelog](https://cljdoc.org/d/juji/editscript/0.6.1/doc/changelog?q=timeout#added) where is says `:diff-str` instead of the correct `str-diff`

- With some small values of timeout, the quick diff algorithm (and large enough data structures to trigger a timeout?) causes the error
```
Execution error (IllegalArgumentException) at editscript.diff.quick/diff-vec (quick.cljc:45).
Don't know how to create ISeq from: clojure.lang.Keyword
```
this can be reproduced by adding `:algo :quick` to the [timeout test](https://github.com/juji-io/editscript/blob/master/test/editscript/core_test.cljc#L160)

**Comment by huahaiy** [comment]
Thank you for testing. These issues are addressed in 0.6.2.
